### PR TITLE
Copter: Add pre-arm check for velocity innovation less than 1m/s

### PR DESCRIPTION
**Pull request by arshPratap**
Adds a pre-arm check so that vertical velocity innovation is less than 1m/s, as mentioned here #14881.Thanks !

### CODE REVIEW

File: ArduCopter/AP_Arming.cpp [hunk 1]
```diff
@@ -929,8 +938,24 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
     // disable gps velocity based centrefugal force compensation
     ahrs.set_correct_centrifugal(false);
     hal.util->set_soft_armed(false);
-
+
```
Comment (rishabsingh3003) [inline ArduCopter/AP_Arming.cpp:941]: Please turn on trailing white-space detection on your editor and remove this

File: ArduCopter/AP_Arming.cpp [hunk 2]
```diff
@@ -552,7 +557,11 @@ bool AP_Arming_Copter::mandatory_gps_checks(bool display_failure)
             return false;
         }
     }
-
+    // check for copter velocity innovations
+    if(!vel_innovation_check()) {
```
Comment (rishabsingh3003) [inline ArduCopter/AP_Arming.cpp:561]: nit pick formatting problems: space after "if"

File: ArduCopter/AP_Arming.cpp [hunk 3]
```diff
@@ -445,7 +445,12 @@ bool AP_Arming_Copter::gps_checks(bool display_failure)
         AP_Notify::flags.pre_arm_gps_check = true;
         return true;
     }
-
+    // check for copter velocity innovations 
+    if (!vel_innovation_check()) {
+        check_failed(ARMING_CHECK_GPS,display_failure,"Bad Velocity");
```
Comment (rishabsingh3003) [inline ArduCopter/AP_Arming.cpp:450]:
```suggestion
        check_failed(ARMING_CHECK_GPS, display_failure, "Bad Velocity");
```

File: ArduCopter/AP_Arming.cpp [hunk 4]
```diff
@@ -929,8 +938,24 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
...
+{
+    Vector3f vel,pos,mag;
+    float tas,yaw;
+    if(!copter.ahrs.get_innovations(vel,pos,mag,tas,yaw)) {
+        return false;
+    } else {
+        if(abs(vel.z) <= 0.5) {
+            return true;
+        }
+        else {
```
Comment (rishabsingh3003) [inline ArduCopter/AP_Arming.cpp:957]: put this one line above to match ArduPilot's style standards

File: ArduCopter/AP_Arming.cpp [hunk 5]
```diff
@@ -929,8 +938,24 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
...
     copter.ap.in_arming_delay = false;
 
     return true;
 }
+// check if vertical velocity innovations are less than 1m/s (0.5m/s specifically)
+bool AP_Arming_Copter::vel_innovation_check()
+{
+    Vector3f vel,pos,mag;
+    float tas,yaw;
+    if(!copter.ahrs.get_innovations(vel,pos,mag,tas,yaw)) {
```
Comment (rishabsingh3003) [inline ArduCopter/AP_Arming.cpp:951]:
```suggestion
    if (!copter.ahrs.get_innovations(vel, pos, mag, tas, yaw)) {
```

File: ArduCopter/AP_Arming.cpp [hunk 6]
```diff
@@ -929,8 +938,24 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
...
+        return false;
+    } else {
+        if(abs(vel.z) <= 0.5) {
```
Comment (rishabsingh3003) [inline ArduCopter/AP_Arming.cpp:954]:
```suggestion
        if ( fabsf(vel.z) <= 0.5f) {
```

File: ArduCopter/AP_Arming.h [hunk 7]
```diff
@@ -52,7 +52,7 @@ class AP_Arming_Copter : public AP_Arming
     bool gcs_failsafe_check(bool display_failure);
     bool winch_checks(bool display_failure) const;
     bool alt_checks(bool display_failure);
-
+    bool vel_innovation_check();
```
Comment (rishabsingh3003) [inline ArduCopter/AP_Arming.h:55]: Can we make this method a const?

**Review by rishabsingh3003 — CHANGES_REQUESTED** [review]
Suggestions in the review stand for this PR and all your future work. Nice work :)

File: ArduCopter/AP_Arming.cpp (hunk 1 above)
Comment (arshPratap) [reply ArduCopter/AP_Arming.cpp]: Thanks for this @rishabsingh3003 !

**Comment by arshPratap** [comment]
Thanks a lot @rishabsingh3003 for the suggestions and I will definitely keep them in mind for future PRs.

File: ArduCopter/AP_Arming.cpp [hunk 8]
```diff
@@ -552,7 +557,11 @@ bool AP_Arming_Copter::mandatory_gps_checks(bool display_failure)
             return false;
         }
     }
-
+    // check for copter velocity innovations
+    if (!vel_innovation_check()) {
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:561]:
```suggestion
    if (mode_request_gps && !vel_innovation_check()) {
```

If someone wants to take off in stabilise with a bad velocity - let them.

Frankly I'm really not sure we should make it a mandatory check.  Are we going to save more vehicles from tipping over?  Sure!  If we never let them arm at all we'll save every vehicle....  How many users are we going to frustrate with this check vs vehicles saved, I wonder.

File: ArduCopter/AP_Arming.cpp [hunk 9]
```diff
@@ -934,3 +943,18 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
 
     return true;
 }
+// check if vertical velocity innovations are less than 1m/s (0.5m/s specifically)
+bool AP_Arming_Copter::vel_innovation_check() const
+{
+    Vector3f vel,pos,mag;
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:949]:
```suggestion
    Vector3f vel, pos, mag;
```

File: ArduCopter/AP_Arming.cpp [hunk 10]
```diff
@@ -934,3 +943,18 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
...
+    float tas,yaw;
+    if (!copter.ahrs.get_innovations(vel, pos, mag, tas, yaw)) {
+        return false;
+    } else {
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:953]:
```suggestion
    }
```

(and reindent rest)

File: ArduCopter/AP_Arming.cpp [hunk 11]
```diff
@@ -934,3 +943,18 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
...
+        if ( fabsf(vel.z) <= 0.5f) {
+            return true;
+        } else {
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:956]:
```suggestion
        }
```

and reindent rest etc

File: ArduCopter/AP_Arming.cpp [hunk 12]
```diff
@@ -934,3 +943,18 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
...
 }
+// check if vertical velocity innovations are less than 1m/s (0.5m/s specifically)
+bool AP_Arming_Copter::vel_innovation_check() const
+{
+    Vector3f vel,pos,mag;
+    float tas,yaw;
+    if (!copter.ahrs.get_innovations(vel, pos, mag, tas, yaw)) {
+        return false;
+    } else {
+        if ( fabsf(vel.z) <= 0.5f) {
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:954]: Invert this and return false, returning true at the end of the function instead.

File: ArduCopter/AP_Arming.cpp [hunk 13]
```diff
@@ -445,7 +445,12 @@ bool AP_Arming_Copter::gps_checks(bool display_failure)
         AP_Notify::flags.pre_arm_gps_check = true;
         return true;
     }
-
+    // check for copter velocity innovations
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:448]: Nothing about this is Copter-specific.  QuadPlanes, for example, could also take advantage of this check.

File: ArduCopter/AP_Arming.cpp [hunk 14]
```diff
@@ -934,3 +943,18 @@ bool AP_Arming_Copter::disarm(const AP_Arming::Method method, bool do_disarm_che
 
     return true;
 }
+// check if vertical velocity innovations are less than 1m/s (0.5m/s specifically)
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:946]:
```suggestion
// check if vertical velocity innovations are less than 0.5m/s
```

 (and adjust elsewhere)

**Review by peterbarker — COMMENTED** [review]
I'm afraid that the EKF actually already has code in place to do this check.  Its just not enabled by default.

https://ardupilot.org/copter/docs/parameters.html#ek2-gps-check-gps-preflight-check

Only found that after doing most of a PR - I'm going to leave those comments in the review, but my reading here is that if any change is required it is simply to change the default parameter value to include vertical velocity.

**Comment by peterbarker** [comment]
(innovation checks are here: https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_NavEKF3/AP_NavEKF3_VehicleStatus.cpp#L271)

**Comment by peterbarker** [comment]
(and the relevant parameter is `EK3_VEL_I_GATE`)

**Review by khancyr — COMMENTED** [review]
this have been override by https://github.com/ArduPilot/ardupilot/pull/21551

Thanks for your PR anyways. I will close this. Please open an update version it you still want the check we hard settled velocity limits and we will discuss it.

**Pull request closed without merging**